patches/0013-p256-armv8-optimizations

```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
++    // Note: the value of input params are changed in the following.
++    ldp     $n0,$n1,[$n_in]
++    ldp     $n2,$n3,[$n_in,#16]
++
++    // A = a3..a0 := n
++    mov     $a0, $n0
++    mov     $a1, $n1
++    mov     $a2, $n2
++    mov     $a3, $n3
++
++    // X = x4..x0 := 1
++    mov     $x0, #1
++    eor     $x1, $x1, $x1
++    eor     $x2, $x2, $x2
++    eor     $x3, $x3, $x3
++    eor     $x4, $x4, $x4
++
++    // Y = y4..y0 := 0
++    eor     $y0, $y0, $y0
++    eor     $y1, $y1, $y1
++    eor     $y2, $y2, $y2
++    eor     $y3, $y3, $y3
++    eor     $y4, $y4, $y4
++
++.Lbeeu_loop:
++    // if B == 0, jump to .Lbeeu_loop_end
++    ${\TEST_B_ZERO}
++
++    // 0 < B < |n|,
++    // 0 < A <= |n|,
++    // (1)      X*a  ==  B   (mod |n|),
++    // (2) (-1)*Y*a  ==  A   (mod |n|)
++
++    // Now divide B by the maximum possible power of two in the
++    // integers, and divide X by the same value mod |n|.
++    // When we're done, (1) still holds.
++
++    // shift := number of trailing 0s in $b0
++    // (      = number of leading 0s in $t1; see the "rbit" instruction in TEST_B_ZERO)
++    clz     $shift, $t1
++
++    // If there is no shift, goto shift_A_Y
++    cbz     $shift, .Lbeeu_shift_A_Y
++
++    // Shift B right by "$shift" bits
++    ${\SHIFT256($b0, $b1, $b2, $b3)}
++
++    // Shift X right by "$shift" bits, adding n whenever X becomes odd.
++    // $shift--;
++    // $t0 := 0; needed in the addition to the most significant word in SHIFT1
++    eor     $t0, $t0, $t0
++.Lbeeu_shift_loop_X:
++    ${\SHIFT1($x0, $x1, $x2, $x3, $x4)}
++    subs    $shift, $shift, #1
++    bne     .Lbeeu_shift_loop_X
++
++    // Note: the steps above perform the same sequence as in p256_beeu-x86_64-asm.pl
++    // with the following differences:
++    // - "$shift" is set directly to the number of trailing 0s in B
++    //   (using rbit and clz instructions)
++    // - The loop is only used to call SHIFT1(X)
++    //   and $shift is decreased while executing the X loop.
++    // - SHIFT256(B, $shift) is performed before right-shifting X; they are independent
++
++.Lbeeu_shift_A_Y:
++    // Same for A and Y.
++    // Afterwards, (2) still holds.
++    // Reverse the bit order of $a0
++    // $shift := number of trailing 0s in $a0 (= number of leading 0s in $t1)
++    rbit    $t1, $a0
++    clz     $shift, $t1
++
++    // If there is no shift, goto |B-A|, X+Y update
++    cbz     $shift, .Lbeeu_update_B_X_or_A_Y
++
++    // Shift A right by "$shift" bits
++    ${\SHIFT256($a0, $a1, $a2, $a3)}
++
++    // Shift Y right by "$shift" bits, adding n whenever Y becomes odd.
++    // $shift--;
++    // $t0 := 0; needed in the addition to the most significant word in SHIFT1
++    eor     $t0, $t0, $t0
++.Lbeeu_shift_loop_Y:
++    ${\SHIFT1($y0, $y1, $y2, $y3, $y4)}
++    subs    $shift, $shift, #1
++    bne     .Lbeeu_shift_loop_Y
++
++.Lbeeu_update_B_X_or_A_Y:
++    // Try T := B - A; if cs, continue with B > A (cs: carry set = no borrow)
```

Comment:
That's a good question. I understand if you try to find the inverse of small values like 1 and 2, T will be ~_n_, just slightly smaller.

Reply:
My followup question is then whether this fit in four 64-bit registers taking sign into account?

Reply:
That's a very good question. I think the table on this page https://community.arm.com/developer/ip-products/processors/b/processors-ip-blog/posts/condition-codes-1-condition-flags-and-codes
explains the condition codes well.
For example, let _unsigned_ B and A be
```
B = 1 = 0x0001
A = 15 = 0x1111
Then, 
B - A = 0x0001 + 0x0001
```
There is no space for the sign bit of A, but the processor's ALU, Arithmetic Logic Unit, treats both signed and unsigned numbers the same.
The result will be `0x0010` (which, if there were extra sign bits, should be `0x11110010 = -0x00001110 = -14`).
Our indication that the result is negative is that the carry bit was not set.
If you do 
```
A - B = 0x1111 + 0x1111
```
There will be a carry, which means no-borrow, because A > B.
Note: If we were doing _signed_ arithmetic, i.e. we had the most significant bit as the sign bit, we would need to check the overflow bit as in the table on that page.

Reply:
A 1-2 sentence with this analysis/argument would be great to add here for future reference.

Reply:
I expanded the comments on this section as per your recommendation.